// @sentry/nextjs/cjs/config/wrappers/index.js
Object.defineProperty(exports, '__esModule', { value: true });

var withSentryGetStaticProps = require('./withSentryGetStaticProps.js');
var withSentryServerSideGetInitialProps = require('./withSentryServerSideGetInitialProps.js');
var withSentryServerSideAppGetInitialProps = require('./withSentryServerSideAppGetInitialProps.js');
var withSentryServerSideDocumentGetInitialProps = require('./withSentryServerSideDocumentGetInitialProps.js');
var withSentryServerSideErrorGetInitialProps = require('./withSentryServerSideErrorGetInitialProps.js');
var withSentryGetServerSideProps = require('./withSentryGetServerSideProps.js');



exports.withSentryGetStaticProps = withSentryGetStaticProps.withSentryGetStaticProps;
exports.withSentryServerSideGetInitialProps = withSentryServerSideGetInitialProps.withSentryServerSideGetInitialProps;
exports.withSentryServerSideAppGetInitialProps = withSentryServerSideAppGetInitialProps.withSentryServerSideAppGetInitialProps;
exports.withSentryServerSideDocumentGetInitialProps = withSentryServerSideDocumentGetInitialProps.withSentryServerSideDocumentGetInitialProps;
exports.withSentryServerSideErrorGetInitialProps = withSentryServerSideErrorGetInitialProps.withSentryServerSideErrorGetInitialProps;
exports.withSentryGetServerSideProps = withSentryGetServerSideProps.withSentryGetServerSideProps;
