// @sentry/nextjs/cjs/config/wrappers/wrapperUtils.js
Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@sentry/core');
var node = require('@sentry/node');
var tracing = require('@sentry/tracing');
var utils = require('@sentry/utils');
var domain = require('domain');

/**
 * Grabs a transaction off a Next.js datafetcher request object, if it was previously put there via
 * `setTransactionOnRequest`.
 *
 * @param req The Next.js datafetcher request object
 * @returns the Transaction on the request object if there is one, or `undefined` if the request object didn't have one.
 */
function getTransactionFromRequest(req) {
  return req._sentryTransaction;
}

function setTransactionOnRequest(transaction, req) {
  req._sentryTransaction = transaction;
}

function autoEndTransactionOnResponseEnd(transaction, res) {
  utils.fill(res, 'end', (originalEnd) => {
    return function ( ...endArguments) {
      transaction.finish();
      return originalEnd.call(this, ...endArguments);
    };
  });
}

/**
 * Wraps a function that potentially throws. If it does, the error is passed to `captureException` and rethrown.
 *
 * Note: This function turns the wrapped function into an asynchronous one.
 */
function withErrorInstrumentation(
  origFunction,
) {
  return async function ( ...origFunctionArguments) {
    try {
      return await origFunction.call(this, ...origFunctionArguments);
    } catch (e) {
      // TODO: Extract error logic from `withSentry` in here or create a new wrapper with said logic or something like that.
      core.captureException(e);
      throw e;
    }
  };
}

/**
 * Calls a server-side data fetching function (that takes a `req` and `res` object in its context) with tracing
 * instrumentation. A transaction will be created for the incoming request (if it doesn't already exist) in addition to
 * a span for the wrapped data fetching function.
 *
 * All of the above happens in an isolated domain, meaning all thrown errors will be associated with the correct span.
 *
 * @param origFunction The data fetching method to call.
 * @param origFunctionArguments The arguments to call the data fetching method with.
 * @param req The data fetching function's request object.
 * @param res The data fetching function's response object.
 * @param options Options providing details for the created transaction and span.
 * @returns what the data fetching method call returned.
 */
function callTracedServerSideDataFetcher(
  origFunction,
  origFunctionArguments,
  req,
  res,
  options

,
) {
  return domain.create().bind(async () => {
    let requestTransaction = getTransactionFromRequest(req);

    if (requestTransaction === undefined) {
      var sentryTraceHeader = req.headers['sentry-trace'];
      var rawBaggageString = req.headers && utils.isString(req.headers.baggage) && req.headers.baggage;
      var traceparentData =
        typeof sentryTraceHeader === 'string' ? utils.extractTraceparentData(sentryTraceHeader) : undefined;

      var baggage = utils.parseBaggageSetMutability(rawBaggageString, traceparentData);

      var newTransaction = core.startTransaction({
        op: 'nextjs.data.server',
        name: options.requestedRouteName,
        ...traceparentData,
        metadata: {
          source: 'route',
          baggage,
        },
      });

      requestTransaction = newTransaction;
      autoEndTransactionOnResponseEnd(newTransaction, res);
      setTransactionOnRequest(newTransaction, req);
    }

    var dataFetcherSpan = requestTransaction.startChild({
      op: 'nextjs.data.server',
      description: `${options.dataFetchingMethodName} (${options.dataFetcherRouteName})`,
    });

    var currentScope = core.getCurrentHub().getScope();
    if (currentScope) {
      currentScope.setSpan(dataFetcherSpan);
      currentScope.addEventProcessor(event =>
        node.addRequestDataToEvent(event, req, {
          include: {
            // When the `transaction` option is set to true, it tries to extract a transaction name from the request
            // object. We don't want this since we already have a high-quality transaction name with a parameterized
            // route. Setting `transaction` to `true` will clobber that transaction name.
            transaction: false,
          },
        }),
      );
    }

    try {
      return await origFunction(...origFunctionArguments);
    } finally {
      dataFetcherSpan.finish();
    }
  })();
}

/**
 * Call a data fetcher and trace it. Only traces the function if there is an active transaction on the scope.
 *
 * We only do the following until we move transaction creation into this function: When called, the wrapped function
 * will also update the name of the active transaction with a parameterized route provided via the `options` argument.
 */
async function callDataFetcherTraced(
  origFunction,
  origFunctionArgs,
  options

,
) {
  const { parameterizedRoute, dataFetchingMethodName } = options;

  var transaction = tracing.getActiveTransaction();

  if (!transaction) {
    return origFunction(...origFunctionArgs);
  }

  // TODO: Make sure that the given route matches the name of the active transaction (to prevent background data
  // fetching from switching the name to a completely other route) -- We'll probably switch to creating a transaction
  // right here so making that check will probabably not even be necessary.
  // Logic will be: If there is no active transaction, start one with correct name and source. If there is an active
  // transaction, create a child span with correct name and source.
  transaction.name = parameterizedRoute;
  transaction.metadata.source = 'route';

  // Capture the route, since pre-loading, revalidation, etc might mean that this span may happen during another
  // route's transaction
  var span = transaction.startChild({
    op: 'nextjs.data.server',
    description: `${dataFetchingMethodName} (${parameterizedRoute})`,
  });

  try {
    return await origFunction(...origFunctionArgs);
  } catch (err) {
    if (span) {
      span.finish();
    }

    // TODO Copy more robust error handling over from `withSentry`
    core.captureException(err);
    throw err;
  }
}

exports.callDataFetcherTraced = callDataFetcherTraced;
exports.callTracedServerSideDataFetcher = callTracedServerSideDataFetcher;
exports.getTransactionFromRequest = getTransactionFromRequest;
exports.withErrorInstrumentation = withErrorInstrumentation;
